src/bucket/Bucket.cpp

```diff
@@ -204,16 +204,69 @@ Bucket::getBucketEntry(LedgerKey const& k)
 // do not load shadowed entries. If we don't find the entry, we do not remove it
 // from keys so that it will be searched for again at a lower level.
 void
-Bucket::loadKeys(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
-                 std::vector<LedgerEntry>& result)
+Bucket::loadKeysWithLimits(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
+                           std::vector<LedgerEntry>& result,
+                           UnorderedMap<LedgerKey, UnorderedSet<Hash>>& lkToTx,
+                           UnorderedMap<Hash, uint32_t>& txReadBytes,
+                           UnorderedSet<LedgerKey>& notLoaded)
 {
     ZoneScoped;
+    auto maxReadQuotaForKey = [&](auto const& key) {
+        uint32_t maxReadQuota = 0;
+        if (lkToTx.empty() || txReadBytes.empty() ||
+            /* classic */ lkToTx.count(key) == 0)
+        {
+            return std::numeric_limits<std::uint32_t>::max();
+        }
+        for (auto const& txn : lkToTx[key])
+        {
+            maxReadQuota = std::max(maxReadQuota, txReadBytes[txn]);
+        }
+        return maxReadQuota;
+    };
+
+    auto updateReadQuotaForLK = [&](auto const& key, auto const& entry) {
+        // At this point, the key has been loaded so it is not added to
+        // the notLoaded set, even if all remaining transactions now have
+        // zero read quota.
+        if (lkToTx.empty() || txReadBytes.empty() ||
+            /* classic */ lkToTx.count(key) == 0)
+        {
+            return;
+        }
+        // Update read quota for all txns waiting on this key.
+        auto size = xdr::xdr_size(entry) + kBucketEntrySizeEncodingBytes;
```

Comment:
TTL entries are not counted towards the transaction read byte limit (we generally don't want users to care about them). So we should not count them towards the limits here as well.